content/en/blog/_posts/non-root-containers-and-devices.md

```diff
@@ -0,0 +1,196 @@
+must be provided:
+
+```json
+{
+        "type": "<string>",
+        "path": "<string>",
+        "major": <int64>,
+        "minor": <int64>,
+        "fileMode": <uint32>,
+        "uid": <uint32>,
+        "gid": <uint32>
+},
+```
+
+The CRI container runtimes (containerd, CRI-O) are responsible for obtaining this information
+from the host for each `Device`. By default, the runtimes copy the host device's user and group IDs:
+
+> `uid` (uint32, OPTIONAL) - id of device owner in the container namespace.  
+> `gid` (uint32, OPTIONAL) - id of device group in the container namespace.
+
+Similarly, the runtimes prepare other mandatory `config.json` sections based on the CRI fields,
+including the ones defined in `securityContext`: `runAsUser`/`runAsGroup`, which become part of the POSIX
+platforms user structure via:
+
+> `uid` (int, REQUIRED) specifies the user ID in the container namespace.  
+> `gid` (int, REQUIRED) specifies the group ID in the container namespace.  
+> `additionalGids` (array of ints, OPTIONAL) specifies additional group IDs in the container namespace to be added to the process.
+
+However, the resulting `config.json` triggers a problem when trying to run containers with
+both devices added and with non-root uid/gid set via `runAsUser`/`runAsGroup`: the container user process
+has no permission to use the device even when its group id (gid, copied from host) was permissive to
+non-root groups. This is because the container user does not belong to that host group (e.g., via `additionalGids`).
+
+Being able to run applications that use devices as non-root user is normal and expected to work so that
+the security principles can be met. Therefore, we looked into alternatives to get the gap filled with what the PodSec/CRI/OCI supports today.
+
+## What was done to solve the issue?
+Some readers probably noticed from the problem definition that it would at least be possible to workaround
+the problem by manually adding the device gid(s) to `supplementalGroups`, or in
+the case of just one device, set `runAsGroup` to the device's group id. However, this is problematic because the device gid(s) may have
+
+different values depending on the nodes' distro/version in the cluster:
+
+```bash
+$ docker run -t fedora:34 grep video /etc/group
+video:x:39:
+$ docker run -t ubuntu:21.04 grep video /etc/group
+video:x:44
+```
+
+Which number to choose in your `securityContext`? Also, what if the `runAsGroup`/`runAsUser` values cannot be hard-coded because
+they are automatically assigned during pod admission time via external security policies?
+
+Unlike volumes with `fsGroup`, the devices have no official notion of `deviceGroup`/`deviceUser` that the CRI runtimes (or kubelet)
+would be able to use. We considered using container annotations set by the device plugins (e.g., `io.kubernetes.cri.hostDeviceSupplementalGroup/`) to get custom OCI `config.json` uid/gid values.
```

Comment:
Who is “we” in this context? Intel?

Reply:
@sftim how to solve the problem was a joint effort between all the people in "thanks" section. I just implemented the fix and wrote the blog. I'd like to keep "we" here but perhaps clarify who's we.

Reply:
I changed my mind because it was easy to change this. Does it look better?